test/org/opensolaris/opengrok/index/IndexDatabaseTest.java

```diff
@@ -77,27 +90,85 @@ public static void tearDownClass() throws Exception {
         defs1 = IndexDatabase.getDefinitions(f1);
         assertNotNull(defs1);
-        assertTrue(defs1.hasSymbol("foobar"));
-        assertTrue(defs1.hasSymbol("a"));
+        assertTrue(defs1.hasSymbol("main"));
+        assertTrue(defs1.hasSymbol("argv"));
         assertFalse(defs1.hasSymbol("b"));
-        assertTrue(defs1.hasDefinitionAt("foobar", 1, new String[1]));
+        assertTrue(defs1.hasDefinitionAt("main", 3, new String[1]));
 
         // Test that we get null back if we request definitions for a file
         // that's not in the repository.
-        File f2 = new File(repository.getSourceRoot() + "/c/foobar.d");
+        File f2 = new File(repository.getSourceRoot() + "/git/foobar.d");
         Definitions defs2 = IndexDatabase.getDefinitions(f2);
         assertNull(defs2);
     }
+
+    private void checkDataExistence(String fileName, boolean shouldExist) {
+        RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+
+        for (String dirName : new String[]{"historycache", IndexDatabase.XREF_DIR}) {
+            File dataDir = new File(env.getDataRootFile(), dirName);
+            File dataFile = new File(dataDir, fileName + ".gz");
+
+            if (shouldExist) {
+                Assert.assertTrue("file " + fileName + " not found in " + dirName,
+                    dataFile.exists());
+            } else {
+                Assert.assertFalse("file " + fileName + " found in " + dirName,
+                    dataFile.exists());
+            }
+        }
+    }
+
+    /**
+     * Test removal of IndexDatabase. xrefs and history index entries after
+     * file has been removed from a repository.
+     *
+     * @throws Exception 
+     */
+    @Test
+    public void testCleanupAfterIndexRemoval() throws Exception {
+        RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+        final int origNumFiles;
+
+        String projectName = "git";
+        String ppath = "/" + projectName;
+        Project project = new Project(projectName, ppath);
+        IndexDatabase idb = new IndexDatabase(project);
+        assertNotNull(idb);
+
+        // Note that the file to remove has to be different than the one used
+        // in {@code testGetDefinitions} because it shares the same index
+        // and this test is going to remove the file and therefore related
+        // definitions.
+        String fileName = "header.h";
+        File gitRoot = new File(repository.getSourceRoot(), projectName);
+        Assert.assertTrue(new File(gitRoot, fileName).exists());
+
+        // Check that the file was indexed successfully in terms of generated data.
+        checkDataExistence(projectName + File.separator + fileName, true);
+        origNumFiles = idb.getNumFiles();
+        Assert.assertEquals(6, origNumFiles);
+
+        // Remove the file and reindex using IndexDatabase directly.
+        File file = new File(repository.getSourceRoot(), projectName + File.separator + fileName);
+        file.delete();
```

Comment:
I don't like having actions done in asserts, makes the code less readable I think.